File: Enginelibrary/src/main/java/com/mobeiwsq/enginelibrary/view/utils/KeyBoardUtils.java
package com.mobeiwsq.enginelibrary.view.utils;

import android.app.Activity;
import android.content.Context;
import android.graphics.Rect;
import android.os.Build;
import android.util.DisplayMetrics;
import android.view.*;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

/**
 * @ClassName : KeyBoardUtils
 * @Description : 虚拟键盘相关实体类
 * @Author : mobeiwsq
 * @Date: 2025/3/24  14:39
 */
public class KeyBoardUtils {

    /**
     * 根据用户点击的坐标获取用户在窗口上触摸到的View，判断这个View是否是EditText来判断是否需要隐藏键盘
     *
     * @param window 窗口
     * @param event  用户点击事件
     * @return 是否需要隐藏键盘
     */
    public static boolean isShouldHideInput(Window window, MotionEvent event) {
        if (window == null || event == null) {
            return false;
        }
        if (!isSoftInputShow(window)) {
            return false;
        }
        if (!(window.getCurrentFocus() instanceof EditText)) {
            return false;
        }
        View decorView = window.getDecorView();
        if (decorView instanceof ViewGroup) {
            return findTouchEditText((ViewGroup) decorView, event) == null;
        }
        return false;
    }

    private static View findTouchEditText(ViewGroup viewGroup, MotionEvent event) {
        if (viewGroup == null) {
            return null;
        }
        for (int i = 0; i < viewGroup.getChildCount(); i++) {
            View child = viewGroup.getChildAt(i);
            if (child == null || !child.isShown()) {
                continue;
            }
            if (!isTouchView(child, event)) {
                continue;
            }
            if (child instanceof EditText) {
                return child;
            } else if (child instanceof ViewGroup) {
                return findTouchEditText((ViewGroup) child, event);
            }
        }
        return null;
    }

    /**
     * 判断view是否在触摸区域内
     *
     * @param view  view
     * @param event 点击事件
     * @return view是否在触摸区域内
     */
    private static boolean isTouchView(View view, MotionEvent event) {
        if (view == null || event == null) {
            return false;
        }
        Rect rect = new Rect();
        view.getGlobalVisibleRect(rect);
        return rect.contains((int) event.getX(), (int) event.getY());
    }

    /**
     * 输入键盘是否在显示
     *
     * @param window 应用窗口
     */
    public static boolean isSoftInputShow(Window window) {
        if (window != null && window.getDecorView() instanceof ViewGroup) {
            return isSoftInputShow((ViewGroup) window.getDecorView());
        }
        return false;
    }

    /**
     * 输入键盘是否在显示
     *
     * @param rootView 根布局
     */
    public static boolean isSoftInputShow(ViewGroup rootView) {
        if (rootView == null) {
            return false;
        }
        int viewHeight = rootView.getHeight();
        //获取View可见区域的bottom
        Rect rect = new Rect();
        rootView.getWindowVisibleDisplayFrame(rect);
        int space = viewHeight - rect.bottom - getNavigationBarHeight(rootView.getContext());
        return space > 0;
    }

    /**
     * 获取系统底部导航栏的高度
     *
     * @param context 上下文
     * @return 系统状态栏的高度
     */
    public static int getNavigationBarHeight(Context context) {
        WindowManager windowManager;
        if (context instanceof Activity) {
            windowManager = ((Activity) context).getWindowManager();
        } else {
            windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        }
        if (windowManager == null) {
            return 0;
        }
        Display defaultDisplay = windowManager.getDefaultDisplay();
        DisplayMetrics realDisplayMetrics = new DisplayMetrics();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            defaultDisplay.getRealMetrics(realDisplayMetrics);
        }
        int realHeight = realDisplayMetrics.heightPixels;
        int realWidth = realDisplayMetrics.widthPixels;

        DisplayMetrics displayMetrics = new DisplayMetrics();
        defaultDisplay.getMetrics(displayMetrics);

        int displayHeight = displayMetrics.heightPixels;
        int displayWidth = displayMetrics.widthPixels;

        if (realHeight - displayHeight > 0) {
            return realHeight - displayHeight;
        }
        return Math.max(realWidth - displayWidth, 0);
    }

    /**
     * 动态隐藏软键盘
     *
     * @param view 视图
     */
    public static void hideSoftInput(final View view) {
        if (view == null) {
            return;
        }
        InputMethodManager imm =
                (InputMethodManager) view.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm == null) {
            return;
        }
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

}

File: engine_project/src/main/java/com/mobeiwsq/engine_project/utils/KeyBoardUtils.kt
package com.mobeiwsq.engine_project.utils

import android.app.Activity
import android.view.MotionEvent
import android.view.View
import android.view.Window
import android.view.inputmethod.InputMethodManager
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

/**
* 软键盘相关工具类
*
* @author : mobeiwsq
* @since :  2025/4/10 11:15
*/
object KeyBoardUtils {
    /**
     *处理默认点击事件，默认处理隐藏输入法
     */
    fun onTouchDown(event: MotionEvent,activity: Activity?) {
        activity?.let {
            if (isShouldHideInput(it.window, event)) {
                it.currentFocus?.let { view ->
                    hideSoftInPut(view)
                }

            }
        }
        return
    }

    /**
     * 是否需要隐藏软键盘
     */
    fun isShouldHideInput(window: Window?, event: MotionEvent): Boolean {
        window?.let {
            if (isInSide(it, it.currentFocus, event) || !isSoftInPutDisplayed(window)) {
                return false
            }
            return true
        }
        return false
    }

    /**
     * 隐藏软键盘
     */
    fun hideSoftInPut(currentFocus: View) {
        currentFocus.clearFocus()
        //关闭软键盘
        val imm = currentFocus.context.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(currentFocus.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
    }

    /**
     * 判断点击坐标是否在EditText内部
     */
    private fun isInSide(window: Window, currentFocus: View?, ev: MotionEvent): Boolean {
        currentFocus?.let {
            val location = intArrayOf(0, 0)
            //获取当前EditText坐标
            currentFocus.getLocationInWindow(location)
            //上下左右
            val left = location[0]
            val top = location[1]
            val right = left + currentFocus.width
            val bottom = top + currentFocus.height
            //点击坐标是否在其内部
            return (ev.x >= left && ev.x <= right && ev.y > top && ev.y < bottom)
        }
        return false
    }

    /**
     * 软键盘是否需要显示
     */
    private fun isSoftInPutDisplayed(window: Window): Boolean {
        return ViewCompat.getRootWindowInsets(window.decorView)
            ?.isVisible(WindowInsetsCompat.Type.ime()) ?: false
    }
}